### PR TITLE
4236 python console autocomplete 2016.09.26 rebased

### PR DESCRIPTION
**Pull request by jcfr**

### CODE REVIEW

File: Libs/Widgets/ctkConsole.cpp [hunk 1]
```diff
@@ -232,6 +233,75 @@ bool ctkConsolePrivate::isMoveLeftWithinLine(QKeyEvent* e, QTextCursor::MoveOper
 }
 
 //-----------------------------------------------------------------------------
+bool ctkConsolePrivate::isMoveRighttWithinLine(QKeyEvent* e, QTextCursor::MoveOperation &moveOperation, QTextCursor::MoveMode &moveMode)
+{
```
Comment (jcfr) [inline Libs/Widgets/ctkConsole.cpp]: @MayeulChassagnard  Move into its own commit

File: Libs/Widgets/ctkConsole.cpp [hunk 2]
```diff
@@ -339,14 +412,42 @@ void ctkConsolePrivate::keyPressEvent(QKeyEvent* e)
...
+    this->updateCompleterIfVisible();
     return;
     }
 
   if (e == QKeySequence::Delete)
     {
+    qDebug() << "Delete";
     e->accept();
-    this->Superclass::keyPressEvent(e);
-    this->updateCommandBuffer();
```
Comment (jcfr) [inline Libs/Widgets/ctkConsole.cpp]: @MayeulChassagnard  Add inline comment justifying this change

File: Libs/Widgets/ctkConsole.cpp [hunk 3]
```diff
@@ -339,14 +412,42 @@ void ctkConsolePrivate::keyPressEvent(QKeyEvent* e)
       this->setTextCursor(text_cursor);
       e->accept();
       }
+    this->updateCompleterIfVisible();
+    return;
+    }
+
+  // End of line should be the end of interactive area
```
Comment (jcfr) [inline Libs/Widgets/ctkConsole.cpp]: @MayeulChassagnard  This block of code should be in its own commit - could be the same as the one adding `isMoveRighttWithinLine`

File: Libs/Widgets/ctkConsole.cpp [hunk 4]
```diff
@@ -313,13 +385,14 @@ void ctkConsolePrivate::keyPressEvent(QKeyEvent* e)
   // Force the cursor back to the interactive area if anything else than copy/paste or page up/down is done
   // but only when a "real" key is pressed, not just a modifier (otherwise we could not press Control-c in the
   // history area because the cursor would jump to the interactive area immediately when Control is pressed)
```
Comment (jcfr) [inline Libs/Widgets/ctkConsole.cpp]: @MayeulChassagnard  update comment to talk about the `message_output_area`

File: Libs/Widgets/ctkConsole.cpp [hunk 5]
```diff
@@ -355,7 +456,11 @@ void ctkConsolePrivate::keyPressEvent(QKeyEvent* e)
   if (e->key() == Qt::Key_Backspace && !(e->modifiers() & ~Qt::ShiftModifier))
     {
     e->accept();
-    if(text_cursor.position() > this->InteractivePosition)
+    // Can delete with backspace when text_cursor is
+    // on InteractivePosition if something selected
```
Comment (jcfr) [inline Libs/Widgets/ctkConsole.cpp]: @MayeulChassagnard Improve the comment

File: Libs/Widgets/ctkConsole.cpp [hunk 6]
```diff
@@ -541,8 +656,72 @@ void ctkConsolePrivate::updateCompleter()
     // Get the text between the current cursor position
     // and the start of the line
     QTextCursor text_cursor = this->textCursor();
-    text_cursor.setPosition(this->InteractivePosition, QTextCursor::KeepAnchor);
+    while (!text_cursor.selectedText().contains(">>>"))
```
Comment (jcfr) [inline Libs/Widgets/ctkConsole.cpp]: @MayeulChassagnard Move this as a last commit labelled `ENH: **WIP** Support completion within multi-line statement`

**Comment by MayeulChassagnard** [comment]
For now, changes are [here](https://github.com/commontk/CTK/compare/master...MayeulChassagnard:4236-python-console-completion-2016.09.27). Let me know if I must push force here or create a new PR

**Comment by jcfr** [comment]
You could push an other branch.

I will review next week when back from travelling.

Thanks
Jc

**Comment by jcfr** [comment]
Closing. Superseded by #688
